### PR TITLE
JSON API

### PR DESCRIPTION
**Pull request by jone**
The JSON API allows to:
- detect Plone sites
- inspect profiles and upgrades for a Plone site
- execute upgrades per Plone site

#### Accessing the API
The API is available on the Zope app root and on the Plone site by the view named `upgrades-api`.
Example: `http://localhost:8080/upgrades-api/list_plone_sites`

#### Authentication
The API requires the `cmf.ManagePortal` permission (`Manager` role). There is not yet any custom authentication mechanism, just use digest authentication.
The plan is to also extend `bin/upgrade`; at that point we can think of a negotiation solution since it is on the same machine.

#### Versioning
The API version can be requested explicitly by adding the version to the URL.
Example: `http://localhost:8080/upgrades-api/v1/list_plone_sites`

#### Error Handling
Known errors are returned as JSON so that a client can read it. Application crashes (unhandled exceptions) still may be returned as HTML.
Example:

```sh
$ curl -uadmin:admin http://localhost:8080/upgrades-api/foo -D -
HTTP/1.1 404 Unkown API action
Server: Zope/(2.13.22, python 2.7.3, darwin) ZServer/1.1
Date: Wed, 14 Jan 2015 10:43:19 GMT
Content-Length: 66
Content-Type: application/json; charset=utf-8

["ERROR", "Unkown API action", "There is no API action \"foo\"."]
```

Some handled errors:
- Wrong API version
- Unkown API action
- Unauthorized
- Method not allowed  (POST / GET)
- Missing required params
- Plone site outdated
- Cyclic dependencies
- Selected upgrade not found

#### Discoverability
When the API is accessed without an action, the description of available actions
is returned.

```sh
$ curl -uadmin:admin http://localhost:8080/upgrades-api
{
    "api_version": "v1",
    "actions": [
        {
            "request_method": "GET",
            "required_params": [],
            "name": "list_plone_sites",
            "description": "Returns a list of Plone sites."
        }
    ]
}
```

```sh
$ curl -uadmin:admin http://localhost:8080/Plone/upgrades-api
{
    "api_version": "v1",
    "actions": [
        {
            "request_method": "POST",
            "required_params": [],
            "name": "execute_proposed_upgrades",
            "description": "Executes all proposed upgrades."
        },
        {
            "request_method": "POST",
            "required_params": [
                "upgrades:list"
            ],
            "name": "execute_upgrades",
            "description": "Executes a list of upgrades, each identified by the upgrade ID in the form \"[dest-version]@[profile ID]\"."
        },
        {
            "request_method": "GET",
            "required_params": [
                "profileid"
            ],
            "name": "get_profile",
            "description": "Returns the profile with the id \"profileid\" as hash."
        },
        {
            "request_method": "GET",
            "required_params": [],
            "name": "list_profiles",
            "description": "Returns a list of all installed profiles."
        },
        {
            "request_method": "GET",
            "required_params": [],
            "name": "list_profiles_proposing_upgrades",
            "description": "Returns a list of profiles with proposed upgrade steps. The upgrade steps of each profile only include proposed upgrades."
        },
        {
            "request_method": "GET",
            "required_params": [],
            "name": "list_proposed_upgrades",
            "description": "Returns a list of proposed upgrades."
        }
    ]
}
```

#### Response types

The standard response content type is JSON.
On unexpected application crashes, HTML may be returned (Zope error handling).
When executing upgrades, the plain text log is streamed.

#### Response streams when executing upgrades

When executing upgrades (``execute_*`` actions), the upgrade log is streamed in
the request. This allows the client to show the progress while a longer upgrade
is running.

Streaming the response means that the response headers are sent before the
upgrade result is available, exceptions might happen later.
Therefore an upgrade execution request always has the status `200 OK`.

The installation result is added at the end of the request:

```sh
$ curl -uadmin:admin -X POST http://localhost:8080/Plone/upgrades-api/execute_proposed_upgrades -D -
HTTP/1.1 200 OK
Server: Zope/(2.13.22, python 2.7.3, darwin) ZServer/1.1
Date: Wed, 14 Jan 2015 10:58:16 GMT
Transfer-Encoding: chunked

2015-01-14 11:58:16 INFO ftw.upgrade ______________________________________________________________________
2015-01-14 11:58:16 INFO ftw.upgrade UPGRADE STEP ftw.upgrade:default: Foo.
2015-01-14 11:58:16 INFO GenericSetup.rolemap Role / permission map imported.
2015-01-14 11:58:17 INFO GenericSetup.archetypetool Archetype tool imported.
2015-01-14 11:58:17 INFO ftw.upgrade Ran upgrade step Foo. for profile ftw.upgrade:default
2015-01-14 11:58:17 INFO ftw.upgrade Upgrade step duration: 1 second
Result: SUCCESS
```

```sh
$ curl -uadmin:admin -X POST http://localhost:8080/Plone/upgrades-api/execute_proposed_upgrades -D -
HTTP/1.1 200 OK
Server: Zope/(2.13.22, python 2.7.3, darwin) ZServer/1.1
Date: Wed, 14 Jan 2015 10:59:02 GMT
Transfer-Encoding: chunked

2015-01-14 11:59:02 INFO ftw.upgrade ______________________________________________________________________
2015-01-14 11:59:02 INFO ftw.upgrade UPGRADE STEP ftw.upgrade:default: Bar.
2015-01-14 11:59:02 ERROR ftw.upgrade FAILED
Traceback (most recent call last):
  File "/Users/jone/projects/packages/ftw.upgrade/ftw/upgrade/jsonapi/plonesite.py", line 104, in _install_upgrades
    executioner.install_upgrades_by_api_ids(*api_ids)
  File "/Users/jone/projects/packages/ftw.upgrade/ftw/upgrade/executioner.py", line 45, in install_upgrades_by_api_ids
    return self.install(data)
  File "/Users/jone/projects/packages/ftw.upgrade/ftw/upgrade/executioner.py", line 33, in install
    self._upgrade_profile(profileid, upgradeids)
  File "/Users/jone/projects/packages/ftw.upgrade/ftw/upgrade/executioner.py", line 52, in _upgrade_profile
    last_dest_version = self._do_upgrade(profileid, upgradeid) \
  File "/Users/jone/projects/packages/ftw.upgrade/ftw/upgrade/executioner.py", line 67, in _do_upgrade
    step.doStep(self.portal_setup)
  File "/Users/jone/projects/python/cache/eggs/Products.GenericSetup-1.7.5-py2.7.egg/Products/GenericSetup/upgrade.py", line 140, in doStep
    self.handler(tool)
  File "/Users/jone/projects/packages/ftw.upgrade/ftw/upgrade/directory/wrapper.py", line 10, in upgrade_step_wrapper
    result = handler(portal_setup, upgrade_profile)
  File "/Users/jone/projects/packages/ftw.upgrade/ftw/upgrade/step.py", line 34, in __new__
    return obj()
  File "/Users/jone/projects/packages/ftw.upgrade/ftw/upgrade/upgrades/20150114104529_bar/upgrade.py", line 9, in __call__
    raise Exception('Bar')
Exception: Bar
Result: FAILURE
```

#### Changes on existing code

- Gatherer:
  - `get_upgrades` is now deprecated and will be replaced with `get_profiles`, since
    it actually is returning profiles, not upgrades.
  - Upgrade info dicts are extended with `api_id` (e.g. `3006@my.package:default`) and
    the `profileid`.
  - `get_upgrades_by_api_ids` was added, returning upgrades identified by the api id.
- Executioner
  - `install_upgrades_by_api_ids` was added, allowing to install
    a selection of upgrades identified by API ugprade ids.

/cc @maethu @lukasgraf @buchi @deiferni

### CODE REVIEW

File: docs/HISTORY.txt [hunk 1]
```diff
@@ -5,7 +5,20 @@ Changelog
 1.11.1 (unreleased)
 -------------------
 
-- Nothing changed yet.
+- New JSON API implemented, accessible with `/upgrades-api`.
+  [jone]
+
+- Executioner: `install_upgrades_by_api_ids` was added, allowing to install
```
Comment (deiferni) [inline docs/HISTORY.txt]: this should be called `Executor` - hopefully we don't carry out death sentences :wink:

**Comment by deiferni** [comment]
i think this looks :+1:
:green_heart: :yellow_heart: :blue_heart: for a versioned API!

File: docs/HISTORY.txt (hunk 1 above)
Comment (jone) [inline docs/HISTORY.txt]: yeah, but it's an already existing component ...

Comment (deiferni) [inline docs/HISTORY.txt]: true dat - renaming separately is ok as well.

**Comment by deiferni** [comment]
not much going on here - :shipit:

**Comment by jone** [comment]
yay, thanks :wink:

File: docs/HISTORY.txt (hunk 1 above)
Comment (lukasgraf) [inline docs/HISTORY.txt]: @deiferni http://www.cs.uni.edu/~mccormic/humor.html :grin:

**Pull request merged**